keras_nlp/models/xlnet/xlnet_backbone.py

```diff
@@ -0,0 +1,241 @@
+            `[batch_size, sequence_length]`.
+        segment_ids: Segment token indices to indicate first and second portions
+            of the inputs of shape `[batch_size, sequence_length]`.
+        padding_mask: Mask to avoid performing attention on padding token indices
+            of shape `[batch_size, sequence_length]`.
+
+    Returns:
+        last_hidden_state: last hidden state of query state of shape
+            `[batch_size, num_predict, hidden_dim]` if query state is not None
+            otherwise last hidden state of content of shape
+            `[batch_size, sequence_length, hidden_dim]`.
+        new_mems: new memory units returned by the model. These are the conatenated
+            tensors of previous mems and hidden states of most recent pass.
+
+    Examples:
+    ```python
+    import numpy as np
+    from keras_nlp.models import XLNetBackbone
+
+    input_data = {
+        "token_ids": np.array(
+            [460, 5272, 1758, 4905, 9, 4, 3], shape=(1, 7),
+        ),
+        "segment_ids": np.array(
+            [0, 0, 0, 0, 0, 0, 2], shape=(1, 7),
+        ),
+        "padding_mask": np.array(
+            [1, 1, 1, 1, 1, 1, 1], shape=(1, 7)
+        ),
+    }
+
+    # Randomly initialized XLNet encoder with a custom config
+    model = keras_nlp.models.XLNetBackbone(
+        vocabulary_size=32000,
+        num_layers=12,
+        num_heads=12,
+        hidden_dim=768,
+        intermediate_dim=3072,
+    )
+    output = model(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        hidden_dim,
+        intermediate_dim,
+        dropout=0.0,
+        activation="gelu",
+        kernel_initializer_range=0.02,
+        bias_initializer="zeros",
+        layer_norm_epsilon=1e-12,
+        **kwargs,
+    ):
+        # Inputs
+        token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="token_ids"
+        )
+        padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="padding_mask"
+        )
+        segment_ids = keras.Input(
+            shape=(None,), dtype="int32", name="segment_ids"
+        )
+
+        # Content and Query Embedding
+        word_emb, pos_emb = ContentAndQueryEmbedding(
+            vocabulary_size=vocabulary_size,
+            hidden_dim=hidden_dim,
+            dropout=dropout,
+            name="content_query_embedding",
+        )(token_id_input=token_id_input)
+
+        # Apply EncoderBlock Preprocessing Layer
+        preprocessing_outputs = XLNetEncoderBlockPreprocessingLayer(
+            hidden_dim=hidden_dim,
+            kernel_initializer_range=kernel_initializer_range,
+            name="encoder_block_preprocess",
+        )(
+            padding_mask=padding_mask,
+            segment_ids=segment_ids,
+        )
+
+        seg_mat, attn_mask_h, attn_mask_g = preprocessing_outputs
```

Comment:
Consider a global rename of `xx_h` -> `xx_content` and `xx_g` -> `xx_query` (if that's the right naming). Might make this a bit more readable.